### PR TITLE
Added Rouge metric

### PR DESCRIPTION
**Pull request by gucifer**
Added a simple Implementation for the ROUGE Metric.

Description:
Simple Implementation of the ROUGE metric (both Rouge-N and Rouge-L).

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by sdesrozis** [comment]
@gucifer thank you ! As discussed, I will review / update this PR on my side. Therefore, we will ask to @vfdev-5 for the final review before merge.

File: ignite/metrics/rouge.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    ) -> None:
+        self._rougetotal: torch.Tensor = torch.tensor(0, device=device, dtype=torch.double)
+        self._num_examples: int = 0
+        self.beta: float = 0.0
+        self.n: int = 1
+        self.beta, self.n, self.rouge_fn, self.aggregate = self._check_parameters(beta, metric, aggregate)
+        self.alpha = 1 / (1 + self.beta ** 2)
+        super(Rouge, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_parameters(self, beta: float, metric: str, aggregate: str) -> Tuple:
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:122]:
```suggestion
    def _check_parameters(self, beta: float, metric: str, aggregate: str) -> Tuple[float, int, Callable, str]:
```
and thus we can possibly remove redefinitions:
```python
        self.beta: float = 0.0
        self.n: int = 1
```

**Review by vfdev-5 — APPROVED** [review]
@sdesrozis and @gucifer, looks very nice now, thanks !
Nit picking comment

File: ignite/metrics/rouge.py (hunk 1 above)
Comment (gucifer) [reply ignite/metrics/rouge.py]: Fixed it!

**Comment by sdesrozis** [comment]
@vfdev-5 Currently I'm refactoring to have a corpus version. I will push it during the week-end.

**Comment by sdesrozis** [comment]
@gucifer I worked on the PR today. Please, could you review ? Currently, it matches the results of the official perl ROUGE 1.5.5 scripts for single sentence. It almost works on multiple candidates but the average in the perl do resampling that slightly changes the results... I wrote a distributed test in which the process don't use the same data. Therefore, parallel reduction are validated.

Let's see if all works. Maybe it will need some fixes. 

@vfdev-5 Could you review the PR ? Sorry for the inconvenience.

File: ignite/metrics/rouge.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        candidate, references = output[0], output[1]
+        multiref_scores = [self.compute_score(candidate=candidate, reference=reference,) for reference in references]
+        score = self._mutliref_reducer(multiref_scores)
+        self._precision += score.precision
+        self._recall += score.recall
+        precision_recall = score.precision * score.recall
+        if precision_recall > 0:  # avoid zero division
+            self._fmeasure += precision_recall / ((1 - self._alpha) * score.precision + self._alpha * score.recall)
+        print("-********", candidate, references)
+        print(score.precision, score.recall, self._fmeasure)
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:201]: Do we need to keep this prints ?

File: ignite/metrics/rouge.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    def reset(self) -> None:
+        for m in self.internal_metrics:
+            m.reset()
+
+    @reinit__is_reduced
+    def update(self, output: Tuple[Sequence[Any], Sequence[Sequence[Any]]]) -> None:
+        for m in self.internal_metrics:
+            m.update(output)
+
+    @sync_all_reduce()
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:432]: I think we do not need to decorate the method if there is nothing to reduce ...

**Review by vfdev-5 — APPROVED** [review]
@sdesrozis looks great!
I left few comments.

File: ignite/metrics/rouge.py (hunk 2 above)
Comment (sdesrozis) [reply ignite/metrics/rouge.py]: Sorry, I would have removed such debug prints...

File: ignite/metrics/rouge.py [hunk 4]
```diff
@@ -399,15 +401,15 @@ class Rouge(Metric):
 
     def __init__(
         self,
-        metrics: Sequence[str] = None,
+        metrics: Optional[Sequence[str]] = None,
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:403]: Can't we set up one by default ?

File: ignite/metrics/rouge.py (hunk 3 above)
Comment (sdesrozis) [reply ignite/metrics/rouge.py:433]: If the method is not decorated, a warning is raised when run in parallel. However, `Rouge` is a simple proxy to `RougeL` and `RougeN` which are parallel. If you have an idea, I take it !

File: ignite/metrics/rouge.py [hunk 5]
```diff
@@ -171,35 +169,36 @@ def __init__(
...
         self._mutliref_reducer = self._get_multiref_reducer()
 
     def _get_multiref_reducer(self) -> MultiRefReducer:
         if self._multiref == "average":
             return MultiRefAverageReducer()
-        if self._multiref == "best":
+        elif self._multiref == "best":
             return MultiRefBestReducer()
+        else:
+            raise ValueError(f"multiref : wrong value (got : {self._multiref })")
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:180]:
```suggestion
            raise ValueError(f"multiref : wrong value (got : {self._multiref})")
```

File: ignite/metrics/rouge.py [hunk 6]
```diff
@@ -171,35 +169,36 @@ def __init__(
         self._multiref = multiref
         valid_multiref = ["best", "average"]
         if self._multiref not in valid_multiref:
-            raise ValueError(f"aggregation must be in {valid_multiref} (got : {self._multiref })")
+            raise ValueError(f"multiref : valid values are {valid_multiref} (got : {self._multiref })")
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:171]:
```suggestion
            raise ValueError(f"multiref : valid values are {valid_multiref} (got : {self._multiref})")
```

File: ignite/metrics/rouge.py (hunk 3 above)
Comment (vfdev-5) [reply ignite/metrics/rouge.py:433]: What kind of warning it is ? I'm thinking about `MetricsLambda.compute` which is not decorated:
https://github.com/pytorch/ignite/blob/cc76de461f63475f3b792c7c109fede95301556e/ignite/metrics/metrics_lambda.py#L88-L92

File: ignite/metrics/rouge.py (hunk 4 above)
Comment (sdesrozis) [reply ignite/metrics/rouge.py:403]: Done.

File: ignite/metrics/rouge.py (hunk 3 above)
Comment (sdesrozis) [reply ignite/metrics/rouge.py:433]: The warning below 

https://github.com/pytorch/ignite/blob/cc76de461f63475f3b792c7c109fede95301556e/ignite/metrics/metric.py#L215

File: ignite/metrics/rouge.py [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+        self.internal_metrics: List[_BaseRouge] = []
+        for m in metrics:
+            if m == "Rouge-L":
+                self.internal_metrics.append(RougeL(multiref, alpha, output_transform, device))
+            else:
+                match = re.match("Rouge-(?P<ngram>[1-9])", m)
+                if not match:
+                    raise ValueError("Rouge metric name must be 'Rouge-L' or 'Rouge-N' for N in [1 ... 9]")
+                n = int(match.group("ngram"))
+                self.internal_metrics.append(RougeN(n, multiref, alpha, output_transform, device))
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:420]: Nit: maybe we could use input metrics as `Sequence[Union[str, int]]` and receive `["L", 1, 2, 4]` and simplify regexp etc ?

Comment (sdesrozis) [reply ignite/metrics/rouge.py:420]:
```
m = Rouge(metrics=["Rouge-1", "Rouge-2", "Rouge-4", "Rouge-L"], multiref="best", alpha=0.5)
```
would become
```
m = Rouge(metrics=[1, 2, 4, "L"], multiref="best", alpha=0.5)
```
I confess that is more compact but less clear to me. You choose :)

Comment (vfdev-5) [reply ignite/metrics/rouge.py:420]: Class `Rouge` is a group of Rouge-* variants. Maybe, it would make also sense to expose others: RougeN, RougeL in `ignite/metrics/__init__.py` ?

I'd code it even:
```python
m = Rouge(variants=[1, 2, 4, "L"], multiref="best", alpha=0.5)
```
But I have not strong opinion about that, just make things "simplier" probably

File: ignite/metrics/rouge.py [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+        return Score(match=match, candidate=candidate, reference=reference)
+
+
+class MultiRefBestReducer(MultiRefReducer):
+    r"""
+    Reducer for selecting the best score
+    """
+
+    def __call__(self, scores: Sequence[Score]) -> Score:
+        return max(scores, key=lambda x: x.recall())
```
Comment (gucifer) [inline ignite/metrics/rouge.py:149]: Why are we using recall for calculating the best?

Comment (gucifer) [reply ignite/metrics/rouge.py:149]: Also, need to add perl as a dependency to run the perl package I think.

Comment (sdesrozis) [reply ignite/metrics/rouge.py:149]: > Why are we using recall for calculating the best?

That is the algorithm in the paper.

Comment (sdesrozis) [reply ignite/metrics/rouge.py:149]: > Also, need to add perl as a dependency to run the perl package I think.
> Check docs here:
> https://github.com/li-plus/rouge-metric
> 
> Also, if you wish to use the perl implementation to compare against I advise against using this module because the perl module is already downloaded and this package will only compare against that fixed version and not any future updates that might happen in the perl module.

Python 3.8.1 failed, others versions seems ok. Last thing to do before merge :)

Comment (gucifer) [reply ignite/metrics/rouge.py:149]: Can you explain why it failed? I thought it failed because of that

Comment (gucifer) [reply ignite/metrics/rouge.py:149]: Also, Looks great! Much cleaner than mine.

Comment (sdesrozis) [reply ignite/metrics/rouge.py:149]: Ok It is windows and MacOS tests.

Many options are possible for us 
* move to a pythonic ROUGE
* don't execute failing tests on such platforms
* install perl on windows and macos

There is a python version of rouge in `metric-rouge`. I tried it but results seemed different than perl rouge... Maybe we could try this further. 

@gucifer what do you think ?

Comment (gucifer) [reply ignite/metrics/rouge.py:149]: Sure! we can just use pythonic rouge. Let me find one which is most inline with Rouge155

**Comment by gucifer** [comment]
@vfdev-5  @sdesrozis  I changed tests to py-rouge. It needs a specific nltk resource to run and we will have to download it everrytime. Can you review it and tell me if further changes are needed?

File: tests/ignite/metrics/test_rouge.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        engine = Engine(update)
+        m = Rouge(variants=[1, 2, "L"], alpha=0.5, device=metric_device)
+        m.attach(engine, "rouge")
+
+        engine.run(data=list(range(size)), max_epochs=1)
+
+        assert "rouge" in engine.state.metrics
+
+        evaluator = pyrouge.Rouge(
+            metrics=["rouge-n", "rouge-l", "rouge-w"],
```
Comment (sdesrozis) [inline tests/ignite/metrics/test_rouge.py:211]:
```suggestion
            metrics=["rouge-n", "rouge-l"],
```

File: tests/ignite/metrics/test_rouge.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
+import os
+
+import nltk
```
Comment (sdesrozis) [inline tests/ignite/metrics/test_rouge.py:3]: Useless ?

**Comment by sdesrozis** [comment]
@gucifer 2 comments and we merge 🎂

File: tests/ignite/metrics/test_rouge.py (hunk 10 above)
Comment (gucifer) [reply tests/ignite/metrics/test_rouge.py:3]: we need to nltk to download punkt which is needed by py-rouge

**Review by sdesrozis — APPROVED** [review]
LGTM !!

**Comment by sdesrozis** [comment]
Bravo @gucifer !!

File: docs/source/metrics.rst [hunk 11]
```diff
@@ -323,6 +323,7 @@ Complete list of metrics
     precision.Precision
     PSNR
     recall.Recall
+    Rouge
```
Comment (vfdev-5) [inline docs/source/metrics.rst:326]: Please, add in another PR:
```
Rouge
RougeL
RougeN
```

File: ignite/metrics/rouge.py [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+            raise ValueError(f"multiref : valid values are {valid_multiref} (got : {self._multiref})")
+        self._mutliref_reducer = self._get_multiref_reducer()
+
+    def _get_multiref_reducer(self) -> MultiRefReducer:
+        if self._multiref == "average":
+            return MultiRefAverageReducer()
+        elif self._multiref == "best":
+            return MultiRefBestReducer()
+        else:
+            raise ValueError(f"multiref : wrong value (got : {self._multiref})")
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:180]: @gucifer please add a test to cover this case in another PR

File: ignite/metrics/rouge.py [hunk 13]
```diff
@@ -0,0 +1,439 @@
+import re
```
Comment (vfdev-5) [inline ignite/metrics/rouge.py:1]: We can remove `import re` now

**Comment by vfdev-5** [comment]
Thanks a lot @sdesrozis and @gucifer for your hard work on this metric !

@gucifer I left few comments to address in a follow-up PR.
